Experimenting with #1467: Support for @JsonUnwrapped in @JsonCreators

fxshlein opened the pull request:
(fixes #1467 -- one of oldest highly-voted-for open issues)

I was more or less just messing around with an approach to solve this for now without the big refactor mentioned in the issue. Perhaps this is already fine? The test I added is green at least, but I have no idea what I might have missed. After all, if an "object with three properties" test covered a considerable amount of jackson's feature set, I wouldn't be using it... 😆

fxshlein commented on src/main/java/com/fasterxml/jackson/databind/deser/impl/UnwrappedPropertyHandler.java:
```diff
@@ -1,62 +1,112 @@
+        return newProps;
+    }
+
+    public boolean isUnwrapped(SettableBeanProperty property) {
+        return this._unwrappedPropertyNames.contains(property.getName());
+    }
```
Would this be sufficient to check for the unwrapped properties? Not sure how different things affect the result of `getName()`

cowtowncoder replied:
In theory names should be `PropertyName` instances (matters with XML, mostly), but internally a few things are still just `String`s so I suspect this is fine.

cowtowncoder replied:
As per my comment higher up, I am not sure this method is actually needed: at least tests included seem to pass just fine without it being called (as well as all other existing tests).

cowtowncoder commented on src/main/java/com/fasterxml/jackson/databind/deser/impl/UnwrappedPropertyHandler.java:
```diff
@@ -1,62 +1,112 @@
 import com.fasterxml.jackson.databind.util.NameTransformer;
 import com.fasterxml.jackson.databind.util.TokenBuffer;
 
+import java.io.IOException;
```
Could you move these back to where they were? (I assume IDE just reordered them automatically).

(we finally have Coding Style Guide -> https://github.com/FasterXML/jackson/blob/master/contribution/jackson-coding-style-guide.md

that explains intended ordering)

cowtowncoder commented on src/main/java/com/fasterxml/jackson/databind/introspect/POJOPropertiesCollector.java:
```diff
@@ -716,6 +717,14 @@ private void _addCreatorParam(Map<String, POJOPropertyBuilder> props,
         PropertyName pn = _annotationIntrospector.findNameForDeserialization(param);
         boolean expl = (pn != null && !pn.isEmpty());
         if (!expl) {
+            boolean unrwapping = _annotationIntrospector.findUnwrappingNameTransformer(param) != null;
```
typo in name :)

cowtowncoder commented on src/test/java/com/fasterxml/jackson/databind/struct/UnwrappedCreatorParam265Test.java:
```diff
@@ -1,101 +0,0 @@
-package com.fasterxml.jackson.databind.struct;
```
Why was this test deleted? Was it just renamed or ... ?

fxshlein replied:
This was the test that verified that error messages were printed when using `@JsonUnwrapped` in a creator, which hopefully shouldn't be needed anymore :)

cowtowncoder replied:
Ah. Yes, that makes sense. :)

cowtowncoder commented on src/test/java/com/fasterxml/jackson/databind/struct/TestUnwrappedWithJsonCreator.java:
```diff
@@ -0,0 +1,128 @@
+/**
+ * Tests to verify [databind#1467].
+ */
+public class TestUnwrappedWithJsonCreator extends BaseMapTest
```
Newer naming convention leads to `UnwrappedWith[Json]CreatorTest` (we haven't renamed all or even most tests but trying to move in that direction).

cowtowncoder commented:
Whoa! Impressive work. I share your suspicion wrt test coverage (lack thereof), but it is what it is.

I'd need to read this through with more thought and focus to see if there's anything specifically that wouldn't work (or be against design). In the meantime I added some notes on code style issues (which are obviously not super important but might as well mention now).

Also, assuming we get through this to merge it, one thing needed (if not already done) is CLA:

https://github.com/FasterXML/jackson/blob/master/contributor-agreement.pdf

which we need before merging the first contribution. It's usually easiest to print, fill & sign, scan/photo, email to `cla` at fasterxml dot com. Only needed once, good for all future contributions.

cowtowncoder commented on src/test-jdk17/java/com/fasterxml/jackson/databind/records/RecordWithJsonUnwrappedTest.java:
```diff
@@ -0,0 +1,32 @@
+    public void testUnwrappedWithRecord() throws Exception
+    {
+        String json = "{\"unrelated\": \"unrelatedValue\", \"property1\": \"value1\", \"property2\": \"value2\"}";
+        RecordWithJsonUnwrapped outer = MAPPER.readValue(json, RecordWithJsonUnwrapped.class);
```
Works, but might make sense to first serialize from instance, then read back (round-trip). To make both directions work (I realize that this PR is just for deserialization but still)

fxshlein replied:
Changed it to a round-trip. Should I also change the tests in `UnwrappedWithCreatorTest`? Since there are already similar tests for serializing regular classes.

fxshlein commented:
@cowtowncoder I sent the signed CLA to `cla@...`. 🙂 

Just to verify, [here](https://github.com/FasterXML/jackson/blob/master/CONTRIBUTING.md#paperwork) it says to email the CLA to `info@...`, are both fine?

cowtowncoder commented on src/main/java/com/fasterxml/jackson/databind/deser/BeanDeserializer.java:
```diff
@@ -855,7 +855,7 @@ protected Object deserializeUsingPropertyBasedWithUnwrapped(JsonParser p, Deseri
                 continue;
             }
-            if (creatorProp != null) {
+            if (creatorProp != null && !_unwrappedPropertyHandler.isUnwrapped(creatorProp)) {
```
I'll have to think about this; ideally such lookup shouldn't be needed (rather, creatorProp would have knowledge of whether it's created via unwrapping).

cowtowncoder replied:
Oddly enough, removing this call makes no difference; all tests pass the same (including Record test).

So is it actually needed? Either we should have a test that fails without it, or it should just be removed, I think.

fxshlein replied:
Yeah, it seems like this makes no sense anymore, I probably implemented it before adapting the rest of the code. The issue this _would_ have solved is if you have a structure like this:
```java
record Outer(@JsonUnwrapped Inner propertyName) {}

record Inner(String propertyName, String otherPropertyName) {}
```
There is a property called `propertyName` inside and outside the unwrapped structure, so when deserializing and encountering the property called "propertyName", I initially thought it would find that property. But as far as this code is concerned, the outer property is called `"@JsonUnwrapped"`, so it should not clash either way. Because of this fake name, the `_unwrappedPropertyNames` set used for `isUnwrapped` actually only contains `"@JsonUnwrapped"`... 😄

cowtowncoder replied:
Right, I thought it might just contain `@JsonUnwrapped` :)

cowtowncoder commented on src/main/java/com/fasterxml/jackson/databind/deser/impl/UnwrappedPropertyHandler.java:
```diff
@@ -15,49 +18,93 @@
+
+    protected final List<SettableBeanProperty> _creatorProperties;
     protected final List<SettableBeanProperty> _properties;
+    protected final Set<String> _unwrappedPropertyNames;
```
... and if `isUnwrapped()` is not actually needed, can drop this `Set` and code that maintains it.

fxshlein replied:
Removed the set and the check!

cowtowncoder commented on src/test/java/com/fasterxml/jackson/databind/struct/UnwrappedWithCreatorTest.java:
```diff
@@ -0,0 +1,123 @@
+public class UnwrappedWithCreatorTest extends BaseMapTest
+{
+	static class ExplicitWithoutName {
+		private final String _unrelated;
```
@fxshlein I changed internal field names to make 100% sure test could not accidentally work due to field name matching constructor property name (since that would actually work despite `final` field declaration, in absence of Creator parameters).

fxshlein commented on src/test/java/com/fasterxml/jackson/databind/struct/UnwrappedPropertyBasedCreatorWithPrefixTest.java:
```diff
@@ -0,0 +1,42 @@
+
+public class UnwrappedPropertyBasedCreatorWithPrefixTest extends BaseMapTest
+{
+    static class Outer {
+        @JsonUnwrapped(prefix = "inner-")
+        Inner inner;
+    }
+
+    static class Inner {
+        private final String _property;
+
+        public Inner(@JsonProperty("property") String property) {
+            _property = property;
+        }
+
+        public String getProperty() {
+            return _property;
+        }
+    }
```
@cowtowncoder While looking at something else, I noticed that you could not do this because the `NameTransformer` used for unwrapping is not applied to `_propertyBasedCreator`, only to `_beanProperties`. This commit seems to fix that: https://github.com/FasterXML/jackson-databind/pull/4271/commits/d65dfacf8d9ec371910db8f230b775e11828b7c1. I guess it broadly fits the scope of the PR, but I can also remove it and make a separate one. In theory its more or less unrelated.

fxshlein commented:
I noticed a minor issue with the placeholder name: If you use two `@JsonUnwrapped` parameters in a creator without explicitly giving it a name `@JsonProperty`, they will be marked as duplicates, as they both get the placeholder:
```
Caused by: java.lang.IllegalArgumentException: Duplicate creator property "@JsonUnwrapped" (index 2 vs 3) for type ...
```

I added a test case here:
https://github.com/FasterXML/jackson-databind/pull/4271/files#diff-1b40e79714a359efac38026f7dfe9dbf45f4def1877d805437ba48955ae9478eR68-R96

From what I can see, there are three solutions to this:
1. Just let it happen, maybe add a special exception message that tells users to simply distinguish the properties using `@JsonProperty`. On one hand this should fit right in, as the user probably has to have `@JsonProperty` on all parameters anyways, but on the other hand it would be kind of weird since the name is never actually used.
2. Change the placeholder to include the index ("unnamed @JsonUnwrapped property at N"). This is the one I implemented now, which makes the test case work. This is nice since the name is more unique, and the names also double as human-readable identifiers in error messages.
3. Do a refactor to allow unwrapped properties to bypass all the places where properties are placed in a map, using their name as a key. I started this here: https://github.com/fxshlein/jackson-databind/commit/3bd10a06548260445cfcde0643f885de5537a09d, but it's already a pretty massive undertaking without even touching everything.

cowtowncoder commented:
To me (2), adding index for bogus name, sounds reasonable at this point. +1.

wingsofovnia commented:
While looking for a stop-gap solution for this in GitHub, I found [this](https://github.com/eclipse-tractusx/bpdm/blob/15b6fbdcdce513374cf7718e704ec33976157382/bpdm-common/src/main/kotlin/org/eclipse/tractusx/bpdm/common/service/DataClassUnwrappedJsonDeserializer.kt#L39) ([usage](https://github.com/eclipse-tractusx/bpdm/blob/15b6fbdcdce513374cf7718e704ec33976157382/bpdm-gate-api/src/main/kotlin/org/eclipse/tractusx/bpdm/gate/api/model/response/SiteGateOutputResponse.kt#L30)) rather straightforward solution for `@JsonUnwrapped` support in Kotlin data classes. Perhaps this might be helpful for you or others who seek a temp solution.

Looking forward to having this merged. Thank you for the contribution!

eranl commented:
@cowtowncoder, what's the status of this PR?

JooHyukKim commented:
@eranl The submitter stop working 6months ago. Unless target issue is highly voted, it would be reasonable to say that there is small chance of merging unless @fxshlein or someone else picks it back up

cowtowncoder commented:
Quick note: this would need to be re-based on 2.18 which may be tricky due to major POJO property introspection rewrite (or might be easy).
If so, could be considered; definitely the problem to solve is worth tackling.

cowtowncoder commented:
Re-based, but now there are conflicts to resolve... not sure I have time to tackle those right now.

cowtowncoder commented:
Ok. Had a look and... yeah, no. These merge conflicts are actually beyond my repairing.
I think I broke this pr, unfortunately; it would need to be re-created starting from 2.18 branch.
The problem is that lots of code was removed from `src/main/java/com/fasterxml/jackson/databind/deser/BasicDeserializerFactory.java` and replaced with additions in `src/main/java/com/fasterxml/jackson/databind/introspect/POJOPropertiesCollector.java`.

Unfortunately those 2 classes were heavily modified by this PR.

But if anyone has time and interest to reconstruct this PR based off 2.18, we could consider it (or, in near future when 2.18 released, against 2.19).

fxshlein commented:
> @eranl The submitter stop working 6months ago. Unless target issue is highly voted, it would be reasonable to say that there is small chance of merging unless @fxshlein or someone else picks it back up

I'm sick at the moment, but maybe I can give it a shot later this week. I was just waiting for the PR to be merged tho, as far as I was concerned nothing was left to do, right? As far as the last comment, I had already implemented the second solution, I was just presenting alternatives. Maybe that caused a misunderstanding, sorry!

fxshlein commented:
Hi @cowtowncoder! Sorry for the long wait, completely forgot about this PR if I'm honest.

I reimplemented everything on top of `1.18`. I can't rename the branch in the PR, so it's still called `1.17`, but it's the latest branch. I'll also deploy the new updated branch from my fork to our internal maven repo and run our tests using that. Those use lots of `@JsonUnwrapped` annotations in previously unsupported places.

fxshlein commented:
Quick update, I tested the updated branch with our internal code (which was based on the old branch) now, and it still works perfectly. I think this should be ready to merge now 🙂

pjfanning commented on src/main/java/com/fasterxml/jackson/databind/deser/SettableBeanProperty.java:
```diff
@@ -584,6 +585,25 @@ public final Object deserializeWith(JsonParser p, DeserializationContext ctxt,
 
+    /**
+     * Returns a copy of this property, unwrapped using the given {@link NameTransformer}.
+     */
```
Can you add `@since 2.18.0`? likewise with any other new public methods.

pjfanning commented on src/main/java/com/fasterxml/jackson/databind/deser/impl/BeanPropertyMap.java:
```diff
@@ -713,25 +713,6 @@ public String toString()
     /**********************************************************
      */
 
-    protected SettableBeanProperty _rename(SettableBeanProperty prop, NameTransformer xf)
```
this breaks backward compatbility - add back the old method and deprecate it

pjfanning replied:
the method can call the new unwrap method, but we need to keep the existing method (its implementation can change - as long as it the change retains the existing behaviour)

fxshlein replied:
good point, readded it!

cowtowncoder replied:
Looking at this again, while I usually agree, in this case it is an internal method called from nowhere else (and `BeanPropertyMap` not extensible via sub-calling). But it's ok to leave it for now, being just 1-liner.

pjfanning commented on src/main/java/com/fasterxml/jackson/databind/deser/impl/UnwrappedPropertyHandler.java:
```diff
@@ -17,51 +18,83 @@
     }
 
-    protected UnwrappedPropertyHandler(List<SettableBeanProperty> props)  {
+    protected UnwrappedPropertyHandler(List<SettableBeanProperty> creatorProps, List<SettableBeanProperty> props) {
```
you need to keep the existing method and deprecate it

pjfanning commented on src/main/java/com/fasterxml/jackson/databind/deser/impl/UnwrappedPropertyHandler.java:
```diff
@@ -17,51 +18,83 @@
         _properties = props;
     }
 
+    public void addCreatorProperty(SettableBeanProperty property) {
```
needs to have javadoc including its `@since 2.18`

pjfanning commented on src/main/java/com/fasterxml/jackson/databind/deser/impl/UnwrappedPropertyHandler.java:
```diff
@@ -17,51 +18,83 @@
+        return newProps;
+    }
+
+    public PropertyValueBuffer processUnwrappedCreatorProperties(
```
javadoc with `@since 2.18`

pjfanning commented on src/main/java/com/fasterxml/jackson/databind/deser/impl/UnwrappedPropertyHandler.java:
```diff
@@ -17,51 +18,83 @@
+    /**
+     * We need a placeholder for creator properties that don't have a name,
+     * but are marked with `@JsonWrapped` annotation.
+     */
```
`@since 2.18`

cowtowncoder commented:
Quick note: due to timing, this won't make it in 2.18, too risky.

fxshlein commented:
haha, I already figured when I saw the 2.18 RC a few days ago. Changed it to 2.19!

fxshlein commented:
@cowtowncoder I changed the base branch to 1.19, seems to work without conflicts. I guess this could be merged now that the 1.19 branch exists, right?

wingsofovnia commented:
Is there any update on this or reason why this isn't getting merged? Looks all green and all feedback addressed.

JooHyukKim commented:
> Is there any update on this or reason why this isn't getting merged? Looks all green and all feedback addressed.

We need proper planning wrt timing on when features get merged.
And this is not the only feature waiting to get merged.
Read some more and you will see Tatu saying...

    Quick note: due to timing, this won't make it in 2.18, too risky.

Not to mention `@since 2.18` is already outdated.

fxshlein commented:
> > Is there any update on this or reason why this isn't getting merged? Looks all green and all feedback addressed.
> 
> We need proper planning wrt timing on when features get merged. And this is not the only feature waiting to get merged. Read some more and you will see Tatu saying...
> 
> ```
> Quick note: due to timing, this won't make it in 2.18, too risky.
> ```
> 
> Not to mention `@since 2.18` is already outdated.

Note that the target branch for this PR is the development branch for `2.19`, and I changed everything to `@since 2.19` a while ago already, so I think the comment regarding `2.18` timing should be addressed. The timing issue was that, when I updated the PR to target `2.18`, the release candidate for `2.18` was already out, at which point it makes no sense to add additional features. I think for `2.19` it should be fine to merge this right?

cowtowncoder commented:
Went over this once, made tiny stylistic changes but want to go over the thing once more and then get merged. My spider sense is tingling wrt this being possibly risky, still, but at the same time all tests, added ones and old, pass -- and this would solve one of only 8 "most-wanted" issues, so clearly many users would like to see it come available.
So will get this merged, just want to be as sure as I can.

cowtowncoder reviewed (approved):
LGTM. Bit worried about possible name clashes, `POJOPropertiesCollector`, but can deal with problems if they arise.

JooHyukKim commented:
Great job @fxshlein !

cowtowncoder commented:
Agreed. Now if I can get this cleanly merged into 3.0 that's the real test... except for one aspect, got it done. So am hopeful still.
